Source/Scene/ModelOutlineGenerator.js

```diff
@@ -0,0 +1,473 @@
+    return false;
+  }
+
+  // Add new buffer to gltf
+  var bufferId =
+    gltf.buffers.push({
+      byteLength: outlineIndexBuffer.byteLength,
+      extras: {
+        _pipeline: {
+          source: outlineIndexBuffer.buffer,
+        },
+      },
+    }) - 1;
+  loadResources.buffers[bufferId] = outlineIndexBuffer;
+
+  // Add new bufferview
+  var bufferViewId =
+    bufferViews.push({
+      buffer: bufferId,
+      byteOffset: 0,
+      byteLength: outlineIndexBuffer.byteLength,
+      target: WebGLConstants.ELEMENT_ARRAY_BUFFER,
+    }) - 1;
+
+  // Add new accessor
+  var accessorId =
+    accessors.push({
+      bufferView: bufferViewId,
+      byteOffset: 0,
+      componentType:
+        outlineIndexBuffer instanceof Uint16Array
+          ? WebGLConstants.UNSIGNED_SHORT
+          : WebGLConstants.UNSIGNED_INT,
+      count: outlineIndexBuffer.length, // start and end for each line
+    }) - 1;
+
+  mesh.primitives[primitiveId].extensions = {
+    CESIUM_primitive_outline: {
+      indices: accessorId,
+    },
+  };
+  gltf.extensionsUsed.push("CESIUM_primitive_outline");
+
+  return true;
+}
+
+/**
+ * Generates a function for getting the attributes of a vertex with a particular
+ * index from a glTF vertex array
+ * @private
+ */
+function generateVertexAttributeGetter(vertexArray, elementsPerVertex) {
+  return function (index) {
+    return [
+      vertexArray[elementsPerVertex * index],
+      vertexArray[elementsPerVertex * index + 1],
+      vertexArray[elementsPerVertex * index + 2],
+    ];
+  };
+}
+
+/**
+ * Adds a single triangle to the directed half edge map.
+ * @param {*} halfEdgeMap
+ * @param {*} firstVertexIndex
+ * @param {*} triangleStartIndex
+ * @param {*} triangleIndices
+ * @param {*} vertexPositionGetter
+ */
+function addTriangleToEdgeGraph(
+  halfEdgeMap,
+  firstVertexIndex,
+  triangleStartIndex, // in indexedTriangle mode, this is an index into the index buffer. otherwise it's an index to the vertex positions
+  triangleIndices,
+  vertexPositionGetter
+) {
+  var vertexIndexA, vertexIndexB, vertexIndexC;
+  // Each vertex in the triangle
+  if (defined(triangleStartIndex) && defined(triangleIndices)) {
+    vertexIndexA = triangleIndices[triangleStartIndex];
+    vertexIndexB = triangleIndices[triangleStartIndex + 1];
+    vertexIndexC = triangleIndices[triangleStartIndex + 2];
+  } else if (defined(firstVertexIndex)) {
+    vertexIndexA = firstVertexIndex;
+    vertexIndexB = firstVertexIndex + 1;
+    vertexIndexC = firstVertexIndex + 2;
+  } else {
+    throw new DeveloperError(
+      "Either firstVertexIndex, or triangleStartIndex and triangleIndices, must be provided."
+    );
+  }
+
+  // Each half edge in the triangle (one in each direction)
+  var edgePairs = [
+    [vertexIndexA, vertexIndexB],
+    [vertexIndexB, vertexIndexC],
+    [vertexIndexC, vertexIndexA],
+    [vertexIndexC, vertexIndexB],
+    [vertexIndexB, vertexIndexA],
+    [vertexIndexA, vertexIndexC],
```

Comment:
> why do we need to store the half edge going in the other direction?

Because if you comment it out then half the outlines disappear! 😂
Ahaha but more seriously, this is a very good point and I should explain why-- in short, because this breaks for non 2-manifold meshes. Check my upcoming explanation in the code.

> this case wouldn't this result in the face being adjacent to itself?

yes. yes it does. but the angle between a face and itself is 0, so this is probably okay

Reply:
Yup ok I get it now, that makes sense